pkg/operator/staticpod/certsyncpod/certsync_controller_test.go

```diff
@@ -0,0 +1,218 @@
+	)
+
+	newRealFS := func() *fileSystem {
+		fs := realFS
+		return &fs
+	}
+
+	checkDirectoryContents := func(t *testing.T, contentDir string, files map[string][]byte) {
+		// Ensure the content directory is in sync.
+		entries, err := os.ReadDir(contentDir)
+		if err != nil {
+			t.Fatalf("Failed to read directory %q: %v", contentDir, err)
+		}
+		writtenData := make(map[string][]byte, len(entries))
+		for _, entry := range entries {
+			info, err := entry.Info()
+			if err != nil {
+				t.Fatalf("Failed to read file information for %q: %v", entry.Name(), err)
+			}
+			if perm := info.Mode().Perm(); perm != 0600 {
+				t.Errorf("Unexpected file permissions for %q: %v", entry.Name(), perm)
+			}
+
+			content, err := os.ReadFile(filepath.Join(contentDir, entry.Name()))
+			if err != nil {
+				t.Fatalf("Failed to read file %q: %v", entry.Name(), err)
+			}
+			writtenData[entry.Name()] = content
+		}
+		if !cmp.Equal(writtenData, files) {
+			t.Errorf("Unexpected directory content:\n%s\n", cmp.Diff(files, writtenData))
+		}
+	}
+
+	ensureParentDirectoryClean := func(t *testing.T, contentDir string) {
+		// Make sure there are no leftovers in the parent directory.
+		parentDir := filepath.Dir(contentDir)
+		parentEntries, err := os.ReadDir(parentDir)
+		if err != nil {
+			t.Fatalf("Failed to read directory %q: %v", parentDir, err)
+		}
+		if n := len(parentEntries); n != 1 {
+			t.Errorf("Unexpected number of entries in directory %q: %d", parentDir, n)
+			for _, entry := range parentEntries {
+				t.Logf("Parent directory entry: %q", entry.Name())
+			}
+		}
+	}
+
+	checkOnSuccess := func(t *testing.T, contentDir string, directoryData, objectData map[string][]byte, errs []error) {
+		if len(errs) > 0 {
+			t.Fatalf("Unexpected errors when writing new data object: %v", errs)
+		}
+
+		checkDirectoryContents(t, contentDir, objectData)
+		ensureParentDirectoryClean(t, contentDir)
+	}
+
+	checkOnError := func(t *testing.T, contentDir string, directoryData, objectData map[string][]byte, errs []error) {
+		if len(errs) == 0 {
+			t.Error("Expected some errors, got none")
+		}
+
+		checkDirectoryContents(t, contentDir, directoryData)
+		ensureParentDirectoryClean(t, contentDir)
+	}
+
+	type testCase struct {
+		name          string
+		newFS         func() *fileSystem
+		directoryData map[string][]byte
+		objectData    map[string][]byte
+		checkState    func(t *testing.T, contentDir string, directoryData, objectData map[string][]byte, errs []error)
+	}
+
+	errorTestCase := func(name string, newFS func() *fileSystem) testCase {
+		return testCase{
+			name:  name,
+			newFS: newFS,
+			directoryData: map[string][]byte{
+				"tls.crt": []byte("TLS cert"),
+				"tls.key": []byte("TLS key"),
+			},
+			objectData: map[string][]byte{
+				"api.crt": []byte("rotated TLS cert"),
+				"api.key": []byte("rotated TLS key"),
+			},
+			checkState: checkOnError,
+		}
+	}
+
+	testCases := []testCase{
+		{
+			name:          "create target directory",
```

Comment:
```suggestion
			name:          "empty target directory",
```

Reply:
Well, actually not such a great idea, then it's not atomic...

Reply:
Hmm, maybe "target dir doesn't exist"?

I usually describe testcase based on the initial situation, not what this test is meant to do. Its a preference, so I'm not insisting.